Compression.md

```diff
@@ -0,0 +1,145 @@
+ - by default the server will configure its compressor and decompressor to use context takeover .i.e. to reuse compression context
+
+###### 3.3.2 `client_max_window_bits` and `server_max_window_bits` 
+These allow the client and server to share the LZ77 sliding window size. The default value is 15 bits which represents a window size of 32768 (2^15). It is necessary that the client's compressor and server's decompressor share the same LZ77 window size. So must it be between the server's compressor and client's decompressor.
+
+In the `Kitura-WebSocket-NIO` implementation:
+- a client can inform the server of it's compressor's LZ77 window size using the `client_max_window_bits` extension parameter. If the parameter has a valid value, the server will configure its decompressor accordingly and send the same extension parameter in the response, indicating an agreement.
+- a client can also request the server to use a particular  LZ77 window size using the `server_max_window_bits` extension parameter. If the value is valid, the server will configure its compressor accordingly and send the same extension parameter in the response, indicating an agreement.
+
+### 4. Developer notes
+1. The `PermessageDeflateCompressor` and `PermessageDeflaterDecompressor` both consolidate multi-frame messages into a single frame. This loss of framing information may not be serious in typical use-cases. But there may be applications where framing information has to be maintained.
+
+2. As mentioned in one of the examples [here](https://tools.ietf.org/html/rfc7692#section-7.1.3), a client may supply fallback negotiation offers, in case negotiation fails. `Kitura-WebSocket-NIO` hasn't implemented this. We make sure every offer goes through.
+
+3. The LZ77 sliding window value must be passed as a negative parameter to deflateInit2()/inflateInit2(). This informs `zlib` that we use raw deflate streams (as against zlib streams that result with sliding window positive values). See [this](https://github.com/IBM-Swift/Kitura-WebSocket-NIO/pull/26/commits/52614a23dbff37db18e2c0fb70e282921c0bb666).
+
+4. There's an open zlib bug with a sliding window size of 8 bits. See [this](https://github.com/madler/zlib/issues/171). There is a workaround for zlib streams. We implement a similar workaround for raw deflate streams [here](https://github.com/IBM-Swift/Kitura-WebSocket-NIO/pull/26/commits/c9d9b44d56de7c398a526b34dacf7b4302045f73).
+
+5. Clients may negotiate for compression but send uncompressed frames. To handle this, just before decompressing, we check the RSV1 bit of the first frame to make sure it belongs to a compressed message.
+
+6. We can merge the two handlers - `PermessageDeflateCompressor` and `PermessageDeflateDecompressor` into a single duplex handler. We can also make the `WebSocketConnection` a duplex handler.
```

Comment:
Do you have an example of what you mean by this?

Reply:
I've added a short definition of a `ChannelDuplexHandler` and also linked to its API doc (which doesn't have anything helpful, actually).